cupy/random/cupy_distributions.cu

```diff
@@ -0,0 +1,252 @@
+#include <curand_kernel.h>
+#include "cupy_distributions.cuh"
+#include <utility>
+#include <stdio.h>
+
+// Add a state of the generator function
+// that can be used to abstract both curand
+// and custom generators
+
+
+struct rk_state {
+
+    __device__ virtual uint32_t rk_int() {
+        return  0;
+    }
+    __device__ virtual double rk_double() {
+        return  0.0;
+    }
+    __device__ virtual double rk_normal() {
+        return  0.0;
+    }
+};
+
+
+struct curand_xor_state: rk_state {
+    // Valid for  XORWOW and MRG32k3a
+    curandState* _state;
+    int _id;
+
+    __device__ curand_xor_state(int id, intptr_t state) {
+        _state = reinterpret_cast<curandState*>(state) + id;
+        _id = id;
+    }
+    __device__ virtual uint32_t rk_int() {
+        return curand(_state);
+    }
+    __device__ virtual double rk_double() {
+        return curand_uniform(_state);
+    }
+    __device__ virtual double rk_normal() {
+        return curand_normal(_state);
+    }
+};
+
+
+// This design is the same as the dtypes one
+template <typename F, typename... Ts>
+void generator_dispatcher(int generator_id, F f, int bpg, int tpb, cudaStream_t stream, Ts&&... args) {
+   switch(generator_id) {
+       case CURAND_XOR_WOW: return f.template operator()<curand_xor_state>(bpg, tpb, stream, std::forward<Ts>(args)...);
```

Comment:
Data comes from python, so we lose it in the inheritance hierarchy of BitGenerator :( and this makes imposible to directly instatiate a template. We have to use an enum like this.